### PR TITLE
[RLP 2] Further RLP clean-up and tests

### PR DESCRIPTION
**Pull request by AlexandraRoatis**
## Description

- Subsequent tests and optimization to the ones in #571 focusing on the `RLP` and `Value` classes.
- **NOTE:** First PR #571 should be merged, then we switch the branch to `dev` before merging this PR. 

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [ ] New feature.
- [x] Enhancement.
- [x] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

- existing test suite and added tests
- commits cherry-picked to master branch with those in #571 and ran sync from Genesis

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [ ] Any dependent changes have been made.

### CODE REVIEW

File: modRlp/src/org/aion/rlp/CompactEncoder.java [hunk 1]
```diff
@@ -113,37 +113,23 @@
         }
         int oddlen = (nibbles.length - terminator) % 2;
         int flag = 2 * terminator + oddlen;
+
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
```
Comment (aionick) [inline modRlp/src/org/aion/rlp/CompactEncoder.java]: I would suggest using `ByteBuffer` here instead, seems to perform much better than a full stream
http://java-performance.info/java-io-bytearrayoutputstream/

File: modRlp/src/org/aion/rlp/Value.java [hunk 2]
```diff
@@ -294,19 +251,12 @@ public boolean isNull() {
...
-        }
-        if (isList() && asList().isEmpty()) {
-            return true;
-        }
-        return isString() && asString().equals("");
-
+        return isNull() // null
+                || isBytes() && asBytes().length == 0 // empty byte array
+                || isList() && asList().isEmpty() // empty list
+                || isString() && asString().equals(""); // empty string
```
Comment (aionick) [inline modRlp/src/org/aion/rlp/Value.java]: very minor, but `equals` could be replaced here with `String`'s `isEmpty` method

File: modRlp/src/org/aion/rlp/RLP.java [hunk 3]
```diff
@@ -379,8 +261,9 @@ private static void fullTraverse(
                     byte[] item = new byte[length];
                     System.arraycopy(msgData, pos + lengthOfLength + 1, item, 0, length);
 
-                    byte[] rlpPrefix = new byte[lengthOfLength + 1];
-                    System.arraycopy(msgData, pos, rlpPrefix, 0, lengthOfLength + 1);
+                    // not used
+                    // byte[] rlpPrefix = new byte[lengthOfLength + 1];
+                    // System.arraycopy(msgData, pos, rlpPrefix, 0, lengthOfLength + 1);
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/RLP.java]: We may want to include it later if for some reason we want to keep track of the prefix as well.

File: modRlp/src/org/aion/rlp/RLP.java [hunk 4]
```diff
@@ -398,8 +281,9 @@ private static void fullTraverse(
                     byte[] item = new byte[length];
                     System.arraycopy(msgData, pos + 1, item, 0, length);
 
-                    byte[] rlpPrefix = new byte[2];
-                    System.arraycopy(msgData, pos, rlpPrefix, 0, 2);
+                    // not used
+                    // byte[] rlpPrefix = new byte[2];
+                    // System.arraycopy(msgData, pos, rlpPrefix, 0, 2);
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/RLP.java]: same as above

File: modRlp/src/org/aion/rlp/RLP.java [hunk 5]
```diff
@@ -528,12 +411,28 @@ private static DecodeResult decodeList(byte[] data, int pos, int prevPos, int le
             if (inputArray.isEmpty()) {
                 return encodeLength(inputArray.size(), OFFSET_SHORT_LIST);
             }
-            byte[] output = ByteUtil.EMPTY_BYTE_ARRAY;
+
+            List<byte[]> obj = new ArrayList<>();
+            int length = 0;
+            byte[] temp;
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/RLP.java]: Yes, it avoids creating an array for each concatenation.

File: modRlp/test/org/aion/rlp/RLPTest.java [hunk 6]
```diff
@@ -379,18 +378,6 @@ public void test10() {
 //                Hex.toHexString(header));
     }
 
-    @Test
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/test/org/aion/rlp/RLPTest.java]: this version of `fullTraverse` was not used in the production code

File: modRlp/src/org/aion/rlp/Value.java [hunk 7]
```diff
@@ -125,7 +118,20 @@ public long asLong() {
...
+        } else if (value instanceof Byte) {
+            return BigInteger.valueOf((Byte) value);
+        } else if (value instanceof Short) {
+            return BigInteger.valueOf((Short) value);
+        } else if (value instanceof Integer) {
+            return BigInteger.valueOf((Integer) value);
+        } else if (value instanceof Long) {
+            return BigInteger.valueOf((Long) value);
+        }
+        return BigInteger.ZERO;
```
[the remark this answers is not in the document]
Comment (AlexandraRoatis) [reply modRlp/src/org/aion/rlp/Value.java]: added java doc in commit ba5c2954639c95eee8258f5d9c4392d1ad147225

**Pull request merged**